pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java

```diff
@@ -278,11 +280,48 @@ public static void validateIngestionConfig(TableConfig tableConfig, @Nullable Sc
                 "Arguments of a transform function '" + arguments + "' cannot contain the destination column '"
                     + columnName + "'");
           }
+          columnToFunctionEvaluator.put(columnName, expressionEvaluator);
+          columnToTransformExpression.put(columnName, transformFunction);
+        }
+        Map<String, Integer> transformFunctionChainDepth = new HashMap<>();
+        for (String column : columnToFunctionEvaluator.keySet()) {
+          if (transformFunctionChainDepth(column, columnToFunctionEvaluator, transformFunctionChainDepth) > 2) {
+            Set<String> derivedArguments = columnToFunctionEvaluator.get(column).getArguments().stream()
+                .filter(a -> transformFunctionChainDepth.get(a) == 2).collect(Collectors.toSet());
+            throw new IllegalStateException(String.format(
+                "Derived columns: [%s] cannot be used as arguments to the transform function: %s of derived column: %s.",
+                derivedArguments, columnToTransformExpression.get(column), column));
+          }
         }
       }
     }
   }
 
+  /**
+   * Returns the depth of chaining in transform functions. Eg:
```

Comment:
This calculation should be based on whether the argument column exist in the schema, instead of on depth only. A column is derived iff all the arguments of its transform function exist in the schema.

E.g.
- If column A not in the schema (only exist in the source data), then `B = f(A)` is not derived, but `C = f(B)` is derived because column B exists in the schema.
- If column A exists in the schema, then `B = f(A)` is derived, and `C = f(B)` is derived column on a derived column, which is invalid

With this logic, if A is in schema, then `B = f(A)` and `C = f(B)` is invalid, and it can cause problem for the segment loader.

Reply:
true.
i'm starting to think, this validation is an overkill. You have already handled in your PR, that if the column is not found, then just generate default. By adding this validation, we will be preventing derived columns even in cases where it could have been handled.

Reply:
I changed my PR to not generating the column when the derived one cannot be created so that user has a chance to fix it without re-downloading the segment. But yes, I think we might not need this validation as segment creation job can handle this properly.